File: app/src/main/java/com/example/zappalex/popularmovies/data/MovieContentProvider.java
package com.example.zappalex.popularmovies.data;

import android.content.ContentProvider;
import android.content.ContentResolver;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

public class MovieContentProvider extends ContentProvider {

    private static final int FAVORITES = 100;
    private static final int FAVORITES_WITH_ID = 101;
    private MovieDbHelper mMovieDbHelper;
    private ContentResolver mContentResolver;


    private static final UriMatcher sUriMatcher = buildUriMatcher();

    private static UriMatcher buildUriMatcher() {
        UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        uriMatcher.addURI(MovieContract.AUTHORITY, MovieContract.ENDPOINT_FAVORITE_MOVIES, FAVORITES);
        uriMatcher.addURI(MovieContract.AUTHORITY, MovieContract.ENDPOINT_FAVORITE_MOVIES + "/#", FAVORITES_WITH_ID);
        return uriMatcher;
    }

    @Override
    public boolean onCreate() {
        Context context = getContext();
        mMovieDbHelper = new MovieDbHelper(context);
        mContentResolver = context.getContentResolver();
        return true;
    }

    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {

        final SQLiteDatabase db = mMovieDbHelper.getReadableDatabase();
        int match = sUriMatcher.match(uri);
        Cursor returnCursor;

        switch (match) {
            case FAVORITES:
                returnCursor = db.query(
                        MovieContract.FavoriteMovieEntry.TABLE_NAME,
                        null,
                        null,
                        null,
                        null,
                        null,
                        null
                );
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }

        if(mContentResolver != null){
            returnCursor.setNotificationUri(mContentResolver, uri);
        }

        return returnCursor;
    }

    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {

        final SQLiteDatabase db = mMovieDbHelper.getWritableDatabase();
        int match = sUriMatcher.match(uri);
        Uri returnUri;

        switch (match) {
            case FAVORITES:
                long id = db.insert(MovieContract.FavoriteMovieEntry.TABLE_NAME, null, values);
                if (id > 0) {
                    returnUri = ContentUris.withAppendedId(MovieContract.FavoriteMovieEntry.CONTENT_URI, id);
                } else {
                    throw new android.database.SQLException("Failed to insert row into " + uri);
                }
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }

        if(mContentResolver != null){
            mContentResolver.notifyChange(uri, null);
        }

        return returnUri;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        final SQLiteDatabase db = mMovieDbHelper.getWritableDatabase();
        int match = sUriMatcher.match(uri);
        int favoritesDeleted;

        switch (match) {
            case FAVORITES_WITH_ID:
                String id = uri.getPathSegments().get(1);
                favoritesDeleted = db.delete(MovieContract.FavoriteMovieEntry.TABLE_NAME, MovieContract.FavoriteMovieEntry.COLUMN_ID + "=?", new String[]{id});
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }

        if (favoritesDeleted != 0 && mContentResolver != null) {
            mContentResolver.notifyChange(uri, null);
        }

        return favoritesDeleted;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        return 0;
    }

    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }
}

File: app/src/main/java/com/example/zappalex/popularmovies/data/MovieContentProvider.kt
package com.example.zappalex.popularmovies.data

import android.content.ContentProvider
import android.content.ContentResolver
import android.content.ContentUris
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.net.Uri

class MovieContentProvider : ContentProvider() {
    private var movieDbHelper: MovieDbHelper? = null
    private var contentResolver: ContentResolver? = null

    override fun onCreate(): Boolean {
        val context = context
        movieDbHelper = MovieDbHelper(context)
        contentResolver = context.contentResolver
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val db = movieDbHelper?.readableDatabase
        val match = sUriMatcher.match(uri)
        val returnCursor: Cursor?

        when (match) {
            FAVORITES -> returnCursor = db?.query(
                    MovieContract.FavoriteMovieEntry.TABLE_NAME, null, null, null, null, null, null
            )
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }

        contentResolver?.let {
            returnCursor?.setNotificationUri(it, uri)
        }

        return returnCursor
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {

        val db = movieDbHelper?.writableDatabase
        val match = sUriMatcher.match(uri)
        val returnUri: Uri?

        when (match) {
            FAVORITES -> {
                val id = db?.insert(MovieContract.FavoriteMovieEntry.TABLE_NAME, null, values) ?: 0
                if (id > 0) {
                    returnUri = ContentUris.withAppendedId(MovieContract.FavoriteMovieEntry.CONTENT_URI, id)
                } else {
                    throw android.database.SQLException("Failed to insert row into $uri")
                }
            }
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }

        contentResolver?.let {
            it.notifyChange(uri, null)
        }

        return returnUri
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val db = movieDbHelper?.writableDatabase
        val match = sUriMatcher.match(uri)
        val favoritesDeleted: Int?

        when (match) {
            FAVORITES_WITH_ID -> {
                val id = uri.pathSegments[1]
                favoritesDeleted = db?.delete(MovieContract.FavoriteMovieEntry.TABLE_NAME, MovieContract.FavoriteMovieEntry.COLUMN_ID + "=?", arrayOf(id))
            }
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }

        contentResolver?.let {
            it.notifyChange(uri, null)
        }

        return favoritesDeleted ?: 0
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        return 0
    }

    override fun getType(uri: Uri): String? {
        return null
    }

    companion object {
        private val FAVORITES = 100
        private val FAVORITES_WITH_ID = 101
        private val sUriMatcher = buildUriMatcher()

        private fun buildUriMatcher(): UriMatcher {
            val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
            uriMatcher.addURI(MovieContract.AUTHORITY, MovieContract.ENDPOINT_FAVORITE_MOVIES, FAVORITES)
            uriMatcher.addURI(MovieContract.AUTHORITY, MovieContract.ENDPOINT_FAVORITE_MOVIES + "/#", FAVORITES_WITH_ID)
            return uriMatcher
        }
    }
}
